docs/testing.md

```diff
@@ -0,0 +1,74 @@
+[Back to README](../README.md)
+
+## Testing
+
+The application currently uses two frameworks for frontend testing: [Jest](https://jestjs.io/) and
+[Cypress](https://www.cypress.io/). There are plans to implement back-end tests in the future.
+
+### Jest Testing
+
+Frontend tests leverage the [Jest framework](https://jestjs.io/).
+
+To run the Jest test suite, execute the command `docker exec -it webpack_watcher npm test`.
+
+To update snapshots, execute `docker exec -it webpack_watcher npm run-script update-snapshot`.
+
+### Cypress Testing
+
+> **Note:** Cypress tests are still present but have not been maintained and may be removed in the future.
+
+Some frontend tests are implemented using the [Cypress framework](https://www.cypress.io/).
+
+For running Cypress tests locally, it is essential that you have a MyLA instance running locally.
+Launch MyLA from the browser, go to the admin view, and add a user called `donald07` with password `root`,
+first name `donald`, and last name `07`.
+Load the latest depersonalized data dump as described in [Getting Started](getting_started.md).
+Then do the following:
+
+1. Install `cypress` and snapshot plugin.
+    ```
+    npm install cypress@4.12.1
+    npm install cypress-plugin-snapshots -S
+    ```
+
+1. Start Cypress.
+    ```
+    npm run cypress:open
+    ```
+
+When running tests, do not use the "All Tests" button due to unsolved issues.
+Run a specific Cypress test if UI changes were made as part of your work.
+If a snapshot fails due to change in the UI,
+try updating the snapshot from the failed test in the Cypress controlled browser.
+
+### Data Validation Testing
+
+Data validation scripts are in `scripts/data_validation` folder.
+To run the data validation scripts, follow the steps below:
+
+1. Copy `env_validation.hjson` file from `env_sample.hjson` file
+    ```sh
+    cp scripts/data_validation/env_sample.hjson scripts/data_validation/env_validation.hjson
+    ```
+
+2. Update the configuration values. Note the hard-coded Michigan specific filters in various query strings.
+Please adjust those values based on your institution's needs.
+
+3. Run the data validation script within the Docker container:
+
+    - Validate queries using UDP BigQuery Expanded Table:
+    This script will validate queries against the UDP BigQuery `expanded` table,
+    and compare whether the data results are identical with those returned from the `events` table queries.
+        ```sh
```

Comment:
We can drop `sh` here and make it simply as `docker exec -it student_dashboard /bin/bash -c "python scripts/data_validation/validate_udp_events_vs_expanded.py"`

Reply:
The reason the `` ```sh `` was there is it's a hint to Github of what type of [code is in the code block](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting) so it color highlights it correctly. This could have been ["sh/bash/shell-script/zsh"](https://github.com/github/linguist/blob/bf853f1c663903e3ee35935189760191f1c45e1c/lib/linguist/languages.yml#L6045).

It's not actually displayed and is useful to add.

Reply:
I'll add them back, I don't think I saw much of a change with these, but I'll keep it if you want.

Reply:
Done. https://github.com/tl-its-umich-edu/my-learning-analytics/pull/1398/commits/0826f71869fb619b6d3c740682df4d0135839673